File: backend/app/core/rule_engine.py
"""
YojanaSetu Deterministic Rule Engine
The Heart of the Platform - 100% Deterministic Rule Evaluation (Zero Hallucination).
Evaluates citizen profiles against scheme criteria using boolean, numerical, and set operators.
"""

from typing import Any, Tuple, List, Dict
from app.models.schemas import CitizenProfile, Rule, Scheme, RuleMatchEvidence


def _normalize_value(val: Any) -> Any:
    """Normalizes strings for case-insensitive and whitespace-trimmed comparisons."""
    if isinstance(val, str):
        return val.strip().lower()
    if isinstance(val, list):
        return [_normalize_value(x) for x in val]
    return val


def get_profile_field_value(profile: CitizenProfile, field_name: str) -> Any:
    """
    Extracts the field value from CitizenProfile with graceful fallbacks.
    Returns None if the field is not present or set to None.
    """
    field_lower = field_name.strip().lower()

    # Direct attribute lookup
    if hasattr(profile, field_lower):
        return getattr(profile, field_lower)

    # Dictionary lookup fallback
    profile_dict = profile.model_dump()
    if field_lower in profile_dict:
        return profile_dict[field_lower]

    return None


def evaluate_single_rule(profile: CitizenProfile, rule: Rule) -> Tuple[bool, Any, str]:
    """
    Evaluates a single rule against a citizen's profile.
    Returns: (is_matched, user_actual_value, condition_summary_string)
    """
    user_val = get_profile_field_value(profile, rule.field)
    operator = rule.operator.upper().strip()
    target_val = rule.value

    # If user value is None for a required rule test
    if user_val is None:
        return False, None, f"{rule.field} is missing"

    norm_user = _normalize_value(user_val)
    norm_target = _normalize_value(target_val)

    try:
        if operator == ">=":
            matched = float(norm_user) >= float(norm_target)
            cond = f">= {target_val}"
        elif operator == "<=":
            matched = float(norm_user) <= float(norm_target)
            cond = f"<= {target_val}"
        elif operator == "==":
            matched = norm_user == norm_target
            cond = f"== {target_val}"
        elif operator == "!=":
            matched = norm_user != norm_target
            cond = f"!= {target_val}"
        elif operator == "IN":
            if isinstance(norm_target, list):
                matched = norm_user in norm_target
            else:
                matched = norm_user == norm_target
            cond = f"IN {target_val}"
        elif operator == "NOT_IN":
            if isinstance(norm_target, list):
                matched = norm_user not in norm_target
            else:
                matched = norm_user != norm_target
            cond = f"NOT IN {target_val}"
        elif operator == "BETWEEN":
            if isinstance(target_val, (list, tuple)) and len(target_val) == 2:
                low, high = float(target_val[0]), float(target_val[1])
                matched = low <= float(norm_user) <= high
                cond = f"BETWEEN {low} and {high}"
            else:
                matched = False
                cond = f"INVALID_RANGE {target_val}"
        else:
            matched = False
            cond = f"UNKNOWN_OPERATOR {operator}"

        return matched, user_val, cond

    except (ValueError, TypeError):
        # Type mismatch (e.g. comparing string to number)
        return False, user_val, f"TYPE_MISMATCH ({rule.field}: {user_val} vs {target_val})"


def evaluate_scheme(profile: CitizenProfile, scheme: Scheme) -> Dict[str, Any]:
    """
    Evaluates all rules for a specific scheme against a citizen profile.
    Returns:
    {
        "scheme": scheme,
        "is_eligible": bool,
        "match_percentage": int (0 to 100),
        "total_rules": int,
        "passed_rules_count": int,
        "rule_results": List[Tuple[Rule, bool, Any, str]]
    }
    """
    if not scheme.rules:
        # If a scheme has no rules, everyone qualifies
        return {
            "scheme": scheme,
            "is_eligible": True,
            "match_percentage": 100,
            "total_rules": 0,
            "passed_rules_count": 0,
            "rule_results": [],
        }

    rule_results = []
    passed_count = 0

    # Also check applicable_state if it's a state scheme
    if scheme.level == "state" and scheme.applicable_state:
        state_rule = Rule(
            field="state",
            operator="==",
            value=scheme.applicable_state,
            description_hi=f"आवेदक {scheme.applicable_state} का मूल निवासी होना चाहिए",
            description_en=f"Applicant must be resident of {scheme.applicable_state}",
        )
        matched, u_val, cond = evaluate_single_rule(profile, state_rule)
        rule_results.append((state_rule, matched, u_val, cond))
        if matched:
            passed_count += 1
    
    for r in scheme.rules:
        # If we already added a state rule above and this rule is also for state, avoid duplicate check
        if r.field == "state" and scheme.level == "state" and scheme.applicable_state:
            continue
        matched, u_val, cond = evaluate_single_rule(profile, r)
        rule_results.append((r, matched, u_val, cond))
        if matched:
            passed_count += 1

    total_rules = len(rule_results)
    is_eligible = (passed_count == total_rules)
    match_percentage = int((passed_count / total_rules) * 100) if total_rules > 0 else 100

    return {
        "scheme": scheme,
        "is_eligible": is_eligible,
        "match_percentage": match_percentage,
        "total_rules": total_rules,
        "passed_rules_count": passed_count,
        "rule_results": rule_results,
    }

File: frontend/src/services/ruleEngine.ts
/**
 * YojanaSetu Client-side Deterministic Rule Engine
 * 100% deterministic rule evaluation in browser with zero hallucination.
 * Directly evaluates citizen profiles against scheme criteria.
 */

import {
  CitizenProfile,
  Rule,
  Scheme,
  RuleMatchEvidence,
  EligibilityResult,
} from "@/types/schema";
import rawSchemes from "@/data/schemes.json";

export const ALL_SCHEMES: Scheme[] = rawSchemes as unknown as Scheme[];

function normalizeValue(val: any): any {
  if (typeof val === "string") {
    return val.trim().toLowerCase();
  }
  if (Array.isArray(val)) {
    return val.map(normalizeValue);
  }
  return val;
}

function getProfileFieldValue(profile: CitizenProfile, fieldName: string): any {
  const fieldLower = fieldName.trim().toLowerCase();
  const obj = profile as any;

  if (fieldLower in obj && obj[fieldLower] !== undefined) {
    return obj[fieldLower];
  }
  return null;
}

function formatValueDisplay(val: any): string {
  if (typeof val === "number") {
    if (val >= 100000) {
      const lakhs = val / 100000;
      return lakhs % 1 !== 0 ? `₹${lakhs.toFixed(2)} लाख` : `₹${Math.round(lakhs)} लाख`;
    }
    if (val >= 1000) {
      return `₹${val.toLocaleString("en-IN")}`;
    }
    return String(val);
  }
  if (Array.isArray(val)) {
    return val.map((x) => String(x).toUpperCase()).join(", ");
  }
  if (typeof val === "boolean") {
    return val ? "हाँ (Yes)" : "नहीं (No)";
  }
  if (val === null || val === undefined) {
    return "लागू नहीं (Not Provided)";
  }
  return String(val);
}

function buildEvidence(rule: Rule, userVal: any, isMatched: boolean): RuleMatchEvidence {
  const targetStr = formatValueDisplay(rule.value);
  const prefixHi = isMatched ? "✓ " : "✗ ";
  const prefixEn = isMatched ? "✓ " : "✗ ";

  let evidenceHi = "";
  let evidenceEn = "";

  if (rule.field === "age") {
    if (isMatched) {
      evidenceHi = `${prefixHi}आपकी आयु ${userVal} वर्ष है (नियम: ${rule.description_hi || `${rule.operator} ${targetStr}`})`;
      evidenceEn = `${prefixEn}Your age is ${userVal} (Required: ${rule.description_en || `${rule.operator} ${targetStr}`})`;
    } else {
      evidenceHi = `${prefixHi}आपकी आयु ${userVal} वर्ष है, जो कि आवश्यक सीमा (${rule.description_hi || `${rule.operator} ${targetStr}`}) में नहीं आती`;
      evidenceEn = `${prefixEn}Your age is ${userVal}, which does not satisfy the requirement (${rule.description_en || `${rule.operator} ${targetStr}`})`;
    }
  } else if (rule.field === "gender") {
    const gText = userVal === "female" ? "महिला (Female)" : userVal === "male" ? "पुरुष (Male)" : String(userVal);
    if (isMatched) {
      evidenceHi = `${prefixHi}लिंग: ${gText} (नियम: ${rule.description_hi || targetStr})`;
      evidenceEn = `${prefixEn}Gender: ${gText} matches criteria (${rule.description_en || targetStr})`;
    } else {
      evidenceHi = `${prefixHi}यह योजना केवल ${rule.description_hi || targetStr} के लिए है (आप: ${gText})`;
      evidenceEn = `${prefixEn}Scheme is intended for ${rule.description_en || targetStr} (You: ${gText})`;
    }
  } else if (rule.field === "annual_income") {
    const incDisp = formatValueDisplay(userVal);
    if (isMatched) {
      evidenceHi = `${prefixHi}पारिवारिक वार्षिक आय ${incDisp} है (सीमा: ${rule.description_hi || `${rule.operator} ${targetStr}`})`;
      evidenceEn = `${prefixEn}Annual income is ${incDisp} (Limit: ${rule.description_en || `${rule.operator} ${targetStr}`})`;
    } else {
      evidenceHi = `${prefixHi}आपकी वार्षिक आय ${incDisp} निर्धारित आय सीमा (${rule.description_hi || `${rule.operator} ${targetStr}`}) से अधिक है`;
      evidenceEn = `${prefixEn}Annual income ${incDisp} exceeds the maximum limit (${rule.description_en || `${rule.operator} ${targetStr}`})`;
    }
  } else if (rule.field === "occupation") {
    if (isMatched) {
      evidenceHi = `${prefixHi}पेशा: ${userVal} (पात्र व्यवसाय: ${rule.description_hi || targetStr})`;
      evidenceEn = `${prefixEn}Occupation: ${userVal} matches required criteria (${rule.description_en || targetStr})`;
    } else {
      evidenceHi = `${prefixHi}यह योजना ${rule.description_hi || targetStr} के लिए है (आपका पेशा: ${userVal})`;
      evidenceEn = `${prefixEn}Scheme requires occupation to be ${rule.description_en || targetStr} (Your occupation: ${userVal})`;
    }
  } else if (rule.field === "state") {
    if (isMatched) {
      evidenceHi = `${prefixHi}राज्य: ${userVal} (योजना क्षेत्र: ${targetStr})`;
      evidenceEn = `${prefixEn}Resident of ${userVal} (Matches scheme territory)`;
    } else {
      evidenceHi = `${prefixHi}यह राज्य योजना केवल ${targetStr} के निवासियों के लिए है (आपका राज्य: ${userVal})`;
      evidenceEn = `${prefixEn}State scheme only applicable to residents of ${targetStr} (Your state: ${userVal})`;
    }
  } else {
    const uStr = formatValueDisplay(userVal);
    if (isMatched) {
      evidenceHi = `${prefixHi}${rule.description_hi || `${rule.field}: ${uStr} (${rule.operator} ${targetStr})`}`;
      evidenceEn = `${prefixEn}${rule.description_en || `${rule.field}: ${uStr} (${rule.operator} ${targetStr})`}`;
    } else {
      evidenceHi = `${prefixHi}${rule.description_hi || `${rule.field}: ${uStr} (${rule.operator} ${targetStr} आवश्यक)`}`;
      evidenceEn = `${prefixEn}${rule.description_en || `${rule.field}: ${uStr} (${rule.operator} ${targetStr} required)`}`;
    }
  }

  return {
    field: rule.field,
    condition: `${rule.operator} ${JSON.stringify(rule.value)}`,
    user_value: userVal,
    matched: isMatched,
    evidence_text_hi: evidenceHi,
    evidence_text_en: evidenceEn,
  };
}

export function evaluateSingleRule(
  profile: CitizenProfile,
  rule: Rule
): { matched: boolean; userVal: any; cond: string } {
  const userVal = getProfileFieldValue(profile, rule.field);
  const operator = (rule.operator || "==").toUpperCase().trim();
  const targetVal = rule.value;

  if (userVal === null || userVal === undefined) {
    return { matched: false, userVal: null, cond: `${rule.field} is missing` };
  }

  const normUser = normalizeValue(userVal);
  const normTarget = normalizeValue(targetVal);

  try {
    let matched = false;
    let cond = "";

    switch (operator) {
      case ">=":
        matched = Number(normUser) >= Number(normTarget);
        cond = `>= ${targetVal}`;
        break;
      case "<=":
        matched = Number(normUser) <= Number(normTarget);
        cond = `<= ${targetVal}`;
        break;
      case ">":
        matched = Number(normUser) > Number(normTarget);
        cond = `> ${targetVal}`;
        break;
      case "<":
        matched = Number(normUser) < Number(normTarget);
        cond = `< ${targetVal}`;
        break;
      case "==":
        matched = normUser === normTarget;
        cond = `== ${targetVal}`;
        break;
      case "!=":
        matched = normUser !== normTarget;
        cond = `!= ${targetVal}`;
        break;
      case "IN":
        if (Array.isArray(normTarget)) {
          matched = normTarget.includes(normUser);
        } else {
          matched = normUser === normTarget;
        }
        cond = `IN ${JSON.stringify(targetVal)}`;
        break;
      case "NOT_IN":
        if (Array.isArray(normTarget)) {
          matched = !normTarget.includes(normUser);
        } else {
          matched = normUser !== normTarget;
        }
        cond = `NOT IN ${JSON.stringify(targetVal)}`;
        break;
      case "BETWEEN":
        if (Array.isArray(targetVal) && targetVal.length === 2) {
          const low = Number(targetVal[0]);
          const high = Number(targetVal[1]);
          matched = Number(normUser) >= low && Number(normUser) <= high;
          cond = `BETWEEN ${low} and ${high}`;
        } else {
          matched = false;
          cond = `INVALID_RANGE ${JSON.stringify(targetVal)}`;
        }
        break;
      default:
        matched = false;
        cond = `UNKNOWN_OPERATOR ${operator}`;
    }

    return { matched, userVal, cond };
  } catch {
    return {
      matched: false,
      userVal,
      cond: `TYPE_MISMATCH (${rule.field}: ${userVal} vs ${targetVal})`,
    };
  }
}

export function evaluateScheme(
  profile: CitizenProfile,
  scheme: Scheme
): EligibilityResult {
  const matchedRules: RuleMatchEvidence[] = [];
  const failingRules: RuleMatchEvidence[] = [];
  const ineligibilityReasonsHi: string[] = [];
  const ineligibilityReasonsEn: string[] = [];

  let totalRulesCount = 0;
  let passedRulesCount = 0;

  // Check state restriction for state schemes
  if (scheme.level === "state" && scheme.applicable_state) {
    totalRulesCount++;
    const stateRule: Rule = {
      field: "state",
      operator: "==",
      value: scheme.applicable_state,
      description_hi: `आवेदक ${scheme.applicable_state} का मूल निवासी होना चाहिए`,
      description_en: `Applicant must be resident of ${scheme.applicable_state}`,
    };
    const { matched, userVal } = evaluateSingleRule(profile, stateRule);
    const evidence = buildEvidence(stateRule, userVal, matched);

    if (matched) {
      passedRulesCount++;
      matchedRules.push(evidence);
    } else {
      failingRules.push(evidence);
      ineligibilityReasonsHi.push(evidence.evidence_text_hi);
      ineligibilityReasonsEn.push(evidence.evidence_text_en);
    }
  }

  // Check scheme rules
  if (scheme.rules && scheme.rules.length > 0) {
    for (const rule of scheme.rules) {
      if (
        rule.field === "state" &&
        scheme.level === "state" &&
        scheme.applicable_state
      ) {
        continue;
      }

      totalRulesCount++;
      const { matched, userVal } = evaluateSingleRule(profile, rule);
      const evidence = buildEvidence(rule, userVal, matched);

      if (matched) {
        passedRulesCount++;
        matchedRules.push(evidence);
      } else {
        failingRules.push(evidence);
        ineligibilityReasonsHi.push(evidence.evidence_text_hi);
        ineligibilityReasonsEn.push(evidence.evidence_text_en);
      }
    }
  }

  const isEligible = failingRules.length === 0;
  const matchPercentage =
    totalRulesCount > 0 ? Math.round((passedRulesCount / totalRulesCount) * 100) : 100;

  return {
    scheme_id: scheme.id,
    scheme_name_hi: scheme.name_hi,
    scheme_name_en: scheme.name_en,
    category: scheme.category,
    benefit_amount_text: scheme.benefit_amount_text,
    benefit_type: scheme.benefit_type,
    official_portal_url: scheme.official_portal_url,
    is_eligible: isEligible,
    match_percentage: matchPercentage,
    matched_rules: matchedRules,
    failing_rules: failingRules,
    ineligibility_reasons_hi: ineligibilityReasonsHi,
    ineligibility_reasons_en: ineligibilityReasonsEn,
    required_documents: scheme.documents || [],
  };
}

export function evaluateAllSchemes(
  profile: CitizenProfile,
  includeIneligible = true
) {
  const eligibleSchemes: EligibilityResult[] = [];
  const ineligibleSchemes: EligibilityResult[] = [];

  for (const scheme of ALL_SCHEMES) {
    const result = evaluateScheme(profile, scheme);
    if (result.is_eligible) {
      eligibleSchemes.push(result);
    } else if (includeIneligible) {
      ineligibleSchemes.push(result);
    }
  }

  // Sort eligible schemes by match percentage descending
  eligibleSchemes.sort((a, b) => b.match_percentage - a.match_percentage);
  ineligibleSchemes.sort((a, b) => b.match_percentage - a.match_percentage);

  return {
    total_schemes_evaluated: ALL_SCHEMES.length,
    eligible_count: eligibleSchemes.length,
    ineligible_count: ineligibleSchemes.length,
    eligible_schemes: eligibleSchemes,
    ineligible_schemes: ineligibleSchemes,
  };
}
